### ISSUE TITLE
[ext_proc] let ext_proc filter follows the grpc spec, wait for trailers before continue the filter chain upon last main stream end_stream=true event.

### ISSUE THREAD

**Issue by stevenzzzz** [envoyproxy/envoy#37088]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: *One line description*

*Description*:
> In Grpc, the trailers carries the grpc-status header terminates a rpc stream. Our current flow ignores the trailers by closeStream and resetStream all together, the later resetStream call would signal the remote server a CANCEL, while clean the sidestream and ignore any possible trailers that might have been sent by ext_proc server.

Instead, for the last de/encode{Headers,data} call, ext_proc filter should fire a half-close (closeStream), and wait until the trailers come back, then call continueProcessing(). 

In PR #37087 we moved one step towards this goal: split half-close from reset(), but we would like to move the half-close earlier to the filter callbacks, before onDestroy gets called. 

>**Note**: The [Envoy_collect tool](https://github.com/envoyproxy/envoy/blob/main/tools/envoy_collect/README.md)
gathers a tarball with debug logs, config and the following admin
endpoints: /stats, /clusters and /server_info. Please note if there are
privacy concerns, sanitize the data prior to sharing the tarball/pasting.

*Admin and Stats Output*:
>Include the admin output for the following endpoints: /stats,
/clusters, /routes, /server_info. For more information, refer to the
[admin endpoint documentation.](https://www.envoyproxy.io/docs/envoy/latest/operations/admin)

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Include the access logs and the Envoy logs.

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Call Stack*:
> If the Envoy binary is crashing, a call stack is **required**.
Please refer to the [Bazel Stack trace documentation](https://github.com/envoyproxy/envoy/tree/main/bazel#stack-trace-symbol-resolution).

**Comment by stevenzzzz** [comment]
hey @tyxia/ @yanavlasov could you put a 
 "no stalebot" mark on this issue?

**Comment by yanjunxiang-google** [comment]
Hi, @stevenzzzz , 
The below comments is problematic to me:
“
Instead, for the last de/encode{Headers,data} call, ext_proc filter should fire a half-close (closeStream), and wait until the trailers come back, then call continueProcessing().
”

If a HTTP request does not have trailers, i.e, if it only contains body, then after ext_proc filter receives the last chunk body response from the ext_proc server (end_of_stream = true in this case), it will terminate the external processing and call continueProcessing().  That's what's the ext_proc filter is designed today.  I am a little bit confusing why ext_proc filter need to wait for a trailers coming back.

**Comment by yanjunxiang-google** [comment]
BTW, the gRPC channel between Envoy and the ext_proc server providing the transportation for the actual HTTP request/response.  The ext_proc filter state machine ties to the actual HTTP request/response contents. It does not depend on the gRPC channel status, i.e, whether the gRPC channel had sent the trailers or not, for its state transitions.

**Comment by stevenzzzz** [comment]
Thanks Yanjun. 

> If a HTTP request does not have trailers, i.e, if it only contains body, then after ext_proc filter receives the last chunk body response from the ext_proc server (end_of_stream = true in this case), it will terminate the external processing and call continueProcessing(). That's what's the ext_proc filter is designed today. I am a little bit confusing why ext_proc filter need to wait for a trailers coming back.

Noteworthy the trailers mentioned is not for the "upstream reqest", but the "side-stream" grpc request's trailers. 
a well behaved Grpc server impl should send trailers to end the request from the Server side.  OTOH, for streaming GRPC, it's client's duty to tell GRPC Server that there is no more data by sending a half-close (empty frame with end_stream=true).

**Comment by stevenzzzz** [comment]
> BTW, the gRPC channel between Envoy and the ext_proc server provides the transportation for the actual HTTP request/response in the main stream. The ext_proc filter state machine ties to the content of the mainstream HTTP request/response. It does not depend on the gRPC channel status, i.e, whether the gRPC channel had sent the trailers or not, for its state transitions.
I am not sure I agree. 
I would expect the grpc-client in ext_proc filter to follow the grpc-spec and be a "good client" as well. 
as a matter of fact, Envoy code is ready for grpc comms.

See Router::Filter::onUpstreamTrailers for example: https://github.com/envoyproxy/envoy/blob/main/source/common/router/router.cc#L1743-L1759
Envoy only inc the success/error counter on trailers frame, if it's a grpc request.

**Comment by htuch** [comment]
+1 on this, I think at the very least it confuses monitoring and debugging, since we would not normally expect to see such a high number of stream resets on the traffic happy path.

**Comment by htuch** [comment]
Another reason this is problematic, FWIW, is that HTTP/2 implementations (e.g. nghttp2) have RST_STREAM flood protections. Excessive RST_STREAM frames can cause connection-level errors and shutdown. Envoy's internal HTTP/2 flood protections are smart enough to avoid doing this for streams that have valid responses on them, but I _think_ nghttp2 just looks at raw number of RST_STREAM frames in a time window.

**Comment by tyxia** [comment]
I agreed with high level direction of splitting the close and reset, as it was internally discussed in load reports issue  

![Image](https://github.com/user-attachments/assets/eee8a043-ad5f-4ab2-912a-d6bf824d8df6)


Just some context, this reset logic was initially introduced in https://github.com/envoyproxy/envoy/pull/18166 and seems to fix the crash issue. I think will need to be careful about fix to ensure no regression.

**Comment by stevenzzzz** [comment]
/assign @yanavlasov 
Assign to Yan++ as he is working on the fix.

**Comment by yanavlasov** [comment]
@stevenzzzz do you need to observe trailers from ext_proc server in the ext_proc filter? I.e. to populate response METADATA?

**Comment by stevenzzzz** [comment]
> [@stevenzzzz](https://github.com/stevenzzzz) do you need to observe trailers from ext_proc server in the ext_proc filter? I.e. to populate response METADATA?

no, at least not for now. The trailer processing override in ext_proc is empty.  

But noteworhty if grpc server returns errors via trailers, the onGrpcError will be called. (for now, this isn't true, since ext_proc filter reset the stream immediately after closeStream).

**Comment by yanavlasov** [comment]
This issue will be closed when the https://github.com/envoyproxy/envoy/blob/4e7d93db9cbb71f7e7e6538d75292d1d75b2f18c/source/common/runtime/runtime_features.cc#L179 is flipped to true by default.

**Comment by dprotaso** [comment]
@yanavlasov is there a tracking issue for changing the default

**Comment by yanavlasov** [comment]
NO, these are automatically created at the time of next Envoy release.
